verilog/CST/verilog_tree_json.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/CST/verilog_tree_json.h"
+
+#include "absl/strings/string_view.h"
+#include "common/text/concrete_syntax_leaf.h"
+#include "common/text/concrete_syntax_tree.h"
+#include "common/text/symbol.h"
+#include "common/text/token_info.h"
+#include "common/util/value_saver.h"
+#include "verilog/CST/verilog_nonterminals.h"  // for NodeEnumToString
+#include "verilog/parser/verilog_parser.h"  // for verilog_symbol_name
+#include "json/json.h"
+
+namespace verilog {
+
+class VerilogTreeToJsonConverter : public verible::SymbolVisitor {
+ public:
+  explicit VerilogTreeToJsonConverter(absl::string_view base);
+
+  void Visit(const verible::SyntaxTreeLeaf&) override;
+  void Visit(const verible::SyntaxTreeNode&) override;
+
+  Json::Value TakeJsonValue() { return std::move(json_); }
+
+ protected:
+  // Range of text spanned by syntax tree, used for offset calculation.
+  const verible::TokenInfo::Context context_;
+
+  // JSON tree root
+  Json::Value json_;
+
+  // Pointer to JSON value of currently visited symbol in its parent's
+  // children list.
+  Json::Value* value_;
+};
+
+VerilogTreeToJsonConverter::VerilogTreeToJsonConverter(absl::string_view base)
+    : context_(verible::TokenInfo::Context(base)),
+    json_(Json::objectValue), value_(&json_) {}
+
+void VerilogTreeToJsonConverter::Visit(const verible::SyntaxTreeLeaf& leaf) {
+  *value_ = Json::objectValue;
+  (*value_)["symbol"] = verilog_symbol_name(leaf.get().token_enum());
+  (*value_)["start"]  = leaf.get().left(context_.base);
+  (*value_)["end"]    = leaf.get().right(context_.base);
+}
```

Comment:
Did you mean to include a copy of `leaf.get().text()` somewhere as well?  (related question from README.md)